libsolidity/parsing/Parser.cpp

```diff
@@ -1986,23 +1985,34 @@ pair<vector<ASTPointer<Expression>>, vector<ASTPointer<ASTString>>> Parser::pars
 	return ret;
 }
 
-optional<string> Parser::findLicenseString()
+optional<string> Parser::findLicenseString(std::vector<ASTPointer<ASTNode>> const& _nodes)
 {
 	// We circumvent the scanner here, because it skips non-docstring comments.
-	static regex const licenseRegex("SPDX-License-Identifier:\\s+([^\\n\\r]+)");
-
+	static regex const licenseRegex("SPDX-License-Identifier:\\s*([^\\n\\r\\s]+)");
+
+	// Remove all parts of the source that where referenced by different AST node locations.
+	// This will leave e.g. "global comments".
+	std::string source = m_scanner->source();
+	std::vector<std::string> toBeRemoved;
+	for (auto& node: _nodes)
+		toBeRemoved.emplace_back(source.substr(node->location().start, node->location().end - node->location().start));
+	for (auto& remove: toBeRemoved)
+		boost::replace_first(source, remove, "");
+
+	std::string result;
 	smatch match;
-	if (regex_search(m_scanner->source(), match, licenseRegex))
+	if (regex_search(source, match, licenseRegex))
 		return boost::trim_copy(string(match[1]));
 
 	parserWarning(
 		0_error,
-		{-1, -1, m_scanner->charStream()},
+		{-1, -1, nullptr},
```

Comment:
Why nullpointer? The warning should reference the source file, otherwise we do not know which source file does not have a license.

Reply:
Ah, yes.